packages/testkit-backend/deno/channel.ts

```diff
@@ -0,0 +1,105 @@
+import { TestkitRequest, TestkitResponse } from "./domain.ts";
+import { iterateReader } from "./deps.ts";
+export interface TestkitClient {
+  id: number;
+  requests: () => AsyncIterable<TestkitRequest>;
+  reply: (response: TestkitResponse) => Promise<void>;
+}
+
+export async function* listen(port: number): AsyncIterable<TestkitClient> {
+  let clientId = 0;
+  const listener = Deno.listen({ port });
+
+  for await (const conn of listener) {
+    const id = clientId++;
+    const { requests, reply } = setupRequestsAndReply(conn)
+    yield { id, requests, reply };
+  }
+}
+
+interface State {
+  finishedReading: boolean
+}
+
+function setupRequestsAndReply (conn: Deno.Conn) {
+  const state = { finishedReading: false }
+  const requests = () => readRequests(conn, state);
+  const reply = createReply(conn, state);
+  
+  return { requests, reply }
+}
+
+async function* readRequests(conn: Deno.Conn, state: State ): AsyncIterable<TestkitRequest> {
+  let inRequest = false;
+  let requestString = "";
+  try {
+    for await (const message of iterateReader(conn)) {
+      const rawTxtMessage = new TextDecoder().decode(message);
+      const lines = rawTxtMessage.split("\n");
+      for (const line of lines) {
+        switch (line) {
+          case "#request begin":
+            if (inRequest) {
+              throw new Error("Already in request");
+            }
+            inRequest = true;
+            break;
+          case "#request end":
+            if (!inRequest) {
+              throw new Error("Not in request");
+            }
+            yield JSON.parse(requestString);
+            inRequest = false;
+            requestString = "";
+            break;
+          case "":
+            // ignore empty lines
+            break;
+          default:
+            if (!inRequest) {
+              throw new Error("Not in request");
+            }
+            requestString += line;
+            break;
+        }
+      }
```

Comment:
Feels like this logic is duplicated here. Should the TestKit protocol ever change, you'll have to change it in multiple places. However, I'm not sure how easy this would be to refactor out (same for the reply method).

Reply:
It is quite difficult because of the way the sockets works in Node and Deno. I prefer to left this refactor to the moment we will have to make changes in the protocol.